File: src/qrs-core-js/src/services/signingService.ts
/**
 * SigningService: issue an SDoc (a signed document) under a TCert.
 *
 * Secret inputs (binding `stripped`) are signed into the COSE external AAD but are
 * NOT stored in the SDoc. At verification time the same secret must be supplied to
 * reconstruct the signed bytes — the comparison is cryptographic and bit-exact.
 */
import { cborEncode } from '../cbor/canonical.js';
import type { PrivateJwk, PublicJwk } from '../crypto/jwk.js';
import { QrsCryptoError, QrsNotFoundError, QrsValidationError } from '../errors.js';
import { toHex } from '../id.js';
import { parseSignedObject, sdocIdOf, tcertNumberOf } from '../signedObject/signedObject.js';
import { isFieldType } from '../signedObject/schemas.js';
import { buildSignedObject, verifyParsedSignedObject } from '../signedObject/signedObject.js';
import type { AlgorithmId, SdocId, TcertId } from '../types.js';
import { isStrippedBinding, type FieldSchema } from '../fields/types.js';
import type { ServiceDeps } from './deps.js';

export interface IssueSdocParams {
  tcertId: TcertId;
  /** Field values keyed by field name. Secrets are prompted/supplied like any other value. */
  values: Record<string, unknown>;
  issuedAt?: number;
  /**
   * SDoc validity is NOT stored separately. If a document needs a validity window,
   * put date/datetime fields in the TCert schema with `verifyRules.expressions`
   * (e.g. `>today()`) — the SDoc then carries just those field values.
   */
}

export interface IssueSdocResult {
  sdocId: SdocId;
  bytes: Uint8Array;
  issuedAt: number;
}

export class SigningService {
  constructor(private readonly deps: ServiceDeps) {}

  async issueSdoc(params: IssueSdocParams): Promise<IssueSdocResult> {
    const tcertBytes = await this.deps.certificateStore.get(params.tcertId);
    if (!tcertBytes) throw new QrsNotFoundError(`TCert not found: ${params.tcertId}`);

    const parsed = parseSignedObject(tcertBytes);
    if (parsed.type !== 'tcert') throw new QrsValidationError('Object is not a TCert');

    const data = parsed.data;
    const keyId = parsed.signerKeyId;
    const certificateNumber = tcertNumberOf(parsed);
    const algorithm = data.algorithm as AlgorithmId;
    const publicJwk = data.publicKey as unknown as PublicJwk;

    // The stored TCert must still verify (defensive).
    const provider = this.deps.cryptoRegistry.get(algorithm);
    if (!(await verifyParsedSignedObject(parsed, provider, publicJwk))) {
      throw new QrsCryptoError('Stored TCert failed self-signature verification');
    }

    // Issue-time validity check: never sign a document that cannot be verified.
    const tcertValidity = data.validity as unknown as { validAfter?: number; validBefore?: number } | undefined;
    const now = params.issuedAt ?? this.deps.clock.now();
    if (tcertValidity) {
      if (tcertValidity.validAfter !== undefined && now < tcertValidity.validAfter) {
        throw new QrsValidationError('TCert is not yet valid');
      }
      if (tcertValidity.validBefore !== undefined && now >= tcertValidity.validBefore) {
        throw new QrsValidationError('TCert has expired');
      }
    }

    const rawSchema = data.schema;
    if (!Array.isArray(rawSchema) || rawSchema.length === 0) {
      throw new QrsValidationError('TCert has no document schema and cannot issue documents');
    }
    const fields = rawSchema as unknown as FieldSchema[];
    // Values are stored as a schema-indexed array (position i matches schema[i]).
    // Field names and labels live only in the TCert — never in the SDoc — keeping
    // the SDoc as small as possible for QR transfer.
    const storedValues: unknown[] = [];
    const secrets: Record<string, string> = {};

    for (let i = 0; i < fields.length; i++) {
      const field = fields[i];
      if (!field) continue;
      if (!isFieldType(field.type)) throw new QrsValidationError(`Unsupported field type: ${field.type}`);
      const engine = this.deps.fieldRegistry.get(field.type);
      let value = params.values[field.name];

      // Auto-fill a declared default (e.g. a hidden datetime field defaulting to now).
      if (value === undefined && field.default !== undefined) {
        value = resolveFieldDefault(field, now);
      }

      if (value === undefined) {
        const required = field.inputRules?.required === true;
        if (required) throw new QrsValidationError(`Missing required value for field '${field.name}'`);
        storedValues.push(null); // optional field omitted
        continue;
      }

      const inputError = engine.validateInput(field, value);
      if (inputError) throw new QrsValidationError(`Field '${field.name}': ${inputError.message}`);

      if (isStrippedBinding(field)) {
        secrets[field.name] = String(value);
        storedValues.push(null);
      } else {
        storedValues.push(engine.encode(field, value));
      }
    }

    const issuedAt = params.issuedAt ?? this.deps.clock.now();
    const sdocData: Record<string, unknown> = {
      issuedAt,
      fields: storedValues,
    };

    const privRec = await this.deps.privateKeyStore.load(keyId);
    if (!privRec) throw new QrsNotFoundError(`Issuer private key not available: ${keyId}`);

    // Secrets are carried in the COSE external AAD (signed but not stored).
    const externalAad = Object.keys(secrets).length > 0 ? cborEncode(secrets) : new Uint8Array(0);
    const keyPair = { algorithm, publicJwk, privateJwk: privRec.privateJwk as PrivateJwk };
    const bytes = await buildSignedObject('sdoc', sdocData, keyPair, provider, externalAad, certificateNumber);
    const sdocId = sdocIdOf(bytes);

    await this.deps.documentStore.save(sdocId, bytes);
    return { sdocId, bytes, issuedAt };
  }
}

/** Resolve a field's declared default into a concrete value at signing time. */
function resolveFieldDefault(field: FieldSchema, now: number): unknown {
  const def = field.default;
  if (def !== null && typeof def === 'object' && (def as { kind?: string }).kind === 'now') {
    const d = new Date(now * 1000);
    const pad = (n: number): string => String(n).padStart(2, '0');
    if (field.type === 'date') {
      return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}`;
    }
    return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}Z`;
  }
  return def;
}

File: qrs-core-python/src/qrs/services/signingService.py
"""SigningService: issue an SDoc (a signed document) under a TCert.

Secret inputs (binding ``stripped``) are signed into the COSE external AAD but are
NOT stored in the SDoc. At verification time the same secret must be supplied to
reconstruct the signed bytes — the comparison is cryptographic and bit-exact.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from ..cbor import cbor_encode
from ..crypto.providers import KeyPairMaterial
from ..date_rules import resolve_field_default
from ..deps import ServiceDeps
from ..envelope import (
    build_signed_object,
    parse_signed_object,
    sdoc_id_of,
    verify_parsed_signed_object,
)
from ..errors import QrsCryptoError, QrsNotFoundError, QrsValidationError
from ..fields import FieldSchema, is_stripped_binding
from ..id import to_hex
from ..signed_object import is_field_type

__all__ = ["SigningService", "IssueSdocParams", "IssueSdocResult"]


@dataclass
class IssueSdocParams:
    tcert_id: str
    values: dict[str, Any]
    issued_at: int | None = None


@dataclass
class IssueSdocResult:
    sdoc_id: str
    bytes: bytes
    issued_at: int


class SigningService:
    def __init__(self, deps: ServiceDeps) -> None:
        self._deps = deps

    async def issue_sdoc(self, params: IssueSdocParams) -> IssueSdocResult:
        tcert_bytes = await self._deps.certificate_store.get(params.tcert_id)
        if not tcert_bytes:
            raise QrsNotFoundError(f"TCert not found: {params.tcert_id}")

        parsed = parse_signed_object(tcert_bytes)
        if parsed.type != "tcert":
            raise QrsValidationError("Object is not a TCert")

        data = parsed.data
        key_id = to_hex(data["keyId"])
        certificate_number = data["certificateNumber"]
        algorithm = data["algorithm"]
        public_jwk = data["publicKey"]

        provider = self._deps.crypto_registry.get(algorithm)
        if not verify_parsed_signed_object(parsed, provider, public_jwk):
            raise QrsCryptoError("Stored TCert failed self-signature verification")

        # Issue-time validity check: never sign a document that cannot be verified.
        tcert_validity = data.get("validity") or {}
        now = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        if tcert_validity:
            if tcert_validity.get("validAfter") is not None and now < tcert_validity["validAfter"]:
                raise QrsValidationError("TCert is not yet valid")
            if tcert_validity.get("validBefore") is not None and now >= tcert_validity["validBefore"]:
                raise QrsValidationError("TCert has expired")

        raw_schema = data.get("schema")
        if not isinstance(raw_schema, list) or len(raw_schema) == 0:
            raise QrsValidationError("TCert has no document schema and cannot issue documents")
        fields = [FieldSchema.from_json(f) for f in raw_schema]

        stored_values: list[Any] = []
        secrets: dict[str, str] = {}

        for field in fields:
            if not is_field_type(field.type):
                raise QrsValidationError(f"Unsupported field type: {field.type}")
            engine = self._deps.field_registry.get(field.type)
            value = params.values.get(field.name)

            # Auto-fill a declared default (e.g. a hidden datetime field defaulting to now).
            if value is None and field.default is not None:
                value = resolve_field_default(field, now)

            if value is None:
                required = bool((field.input_rules or {}).get("required"))
                if required:
                    raise QrsValidationError(f"Missing required value for field '{field.name}'")
                stored_values.append(None)
                continue

            input_error = engine.validate_input(field, value)
            if input_error:
                raise QrsValidationError(f"Field '{field.name}': {input_error.message}")

            if is_stripped_binding(field):
                secrets[field.name] = str(value)
                stored_values.append(None)
            else:
                stored_values.append(engine.encode(field, value))

        issued_at = params.issued_at if params.issued_at is not None else self._deps.clock.now()
        # The TCert linkage (keyId + certificate number) is carried in the COSE
        # protected headers (kid + tcertNumber), NOT in the data — matching the
        # reference implementation and keeping the SDoc minimal for QR transfer.
        sdoc_data: dict[str, Any] = {
            "issuedAt": issued_at,
            "fields": stored_values,
        }

        priv_rec = await self._deps.private_key_store.load(key_id)
        if not priv_rec:
            raise QrsNotFoundError(f"Issuer private key not available: {key_id}")

        external_aad = cbor_encode(secrets) if secrets else b""
        key_pair = KeyPairMaterial(
            algorithm=algorithm,
            public_jwk=public_jwk,
            private_jwk=priv_rec["private_jwk"],
        )
        bytes_out = build_signed_object("sdoc", sdoc_data, key_pair, provider, external_aad, certificate_number)
        sdoc_id = sdoc_id_of(bytes_out)

        await self._deps.document_store.save(sdoc_id, bytes_out)
        return IssueSdocResult(sdoc_id=sdoc_id, bytes=bytes_out, issued_at=issued_at)